File: August2022/C++/UniqueMorseCodeWords.cpp
class Solution {    
public:
    // TC: O(n * m) -> n = words.length, m = words[i].length
    // SC: O(n) 
    int uniqueMorseRepresentations(vector<string>& words) {
        unordered_set<string> set;
        string morseTable[] = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
        
        for (string &word : words) {
            string code = "";
            for (char &c : word) {
                int index = c - 'a';
                code += morseTable[index];
            }
            set.insert(code);
        }

        return set.size();
    }
};

File: August2022/Java/UniqueMorseCodeWords.java

// @saorav21994
// TC : O(wordCount*avg_length_of_each_word)
// SC : O(wordCount)

class Solution {
    public int uniqueMorseRepresentations(String[] words) {
        int len = words.length;
        String [] morse = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
        HashSet<String> res = new HashSet();
        for (int i = 0; i < len; i++) {
            String word = words[i];
            int l = word.length();
            StringBuilder str = new StringBuilder("");
            for (int j = 0; j < l; j++) {
                str.append(morse[(int)word.charAt(j)-97]);
            }
            res.add(String.valueOf(str));
        }
        return res.size();
    }
}
